pkg/image/imagecontentsources/alternativesources.go

```diff
@@ -0,0 +1,173 @@
+	}
+	icspObj, err := runtime.Decode(operatorv1alpha1scheme.Codecs.UniversalDeserializer(), icspData)
+	if err != nil {
+		return fmt.Errorf("error decoding ImageContentSourcePolicy from %s: %v", c.ICSPFile, err)
+	}
+	icsp, ok := icspObj.(*operatorv1alpha1.ImageContentSourcePolicy)
+	if !ok {
+		return fmt.Errorf("could not decode ImageContentSourcePolicy from %s", c.ICSPFile)
+	}
+	c.ICSPList = append(c.ICSPList, *icsp)
+	return nil
+}
+
+// alternativeICSPSources gathers list of DockerImageReference objects from list of ImageContentSourcePolicy objects
+func (a *ICSPSources) alternativeImageSources(imageRef imagereference.DockerImageReference) ([]imagereference.DockerImageReference, error) {
+	var imageSources []imagereference.DockerImageReference
+	for _, icsp := range a.ICSPList {
+		repoDigestMirrors := icsp.Spec.RepositoryDigestMirrors
+		var sourceMatches bool
+		for _, rdm := range repoDigestMirrors {
+			rdmRef, err := imagereference.Parse(rdm.Source)
+			if err != nil {
+				return nil, err
+			}
+			if imageRef.AsRepository() == rdmRef.AsRepository() {
+				klog.V(2).Infof("%v RepositoryDigestMirrors source matches given image", imageRef.AsRepository())
+				sourceMatches = true
+			}
+			for _, m := range rdm.Mirrors {
+				if sourceMatches {
+					klog.V(2).Infof("%v RepositoryDigestMirrors mirror added to potential ImageSourcePrefixes from ImageContentSourcePolicy", m)
+					mRef, err := imagereference.Parse(m)
+					if err != nil {
+						return nil, err
+					}
+					imageSources = append(imageSources, mRef)
+				}
+			}
+		}
+	}
+	uniqueMirrors := make([]imagereference.DockerImageReference, 0, len(imageSources))
+	uniqueMap := make(map[imagereference.DockerImageReference]bool)
+	for _, imageSourceMirror := range imageSources {
+		if _, ok := uniqueMap[imageSourceMirror]; !ok {
+			uniqueMap[imageSourceMirror] = true
+			uniqueMirrors = append(uniqueMirrors, imageSourceMirror)
+		}
+	}
+	// make sure at least 1 imagesource
+	// ie, make sure the image passed is included in image sources
+	// this is so the user-given image ref will be tried
+	if len(imageSources) == 0 {
+		imageSources = append(imageSources, imageRef.AsRepository())
+		return imageSources, nil
+	}
+	klog.V(2).Infof("Found sources: %v for image: %v", uniqueMirrors, imageRef)
+	return uniqueMirrors, nil
+}
+
+type SourceList struct {
+	Sources []string
+}
+
+// Next() satisfies registryclient.ImageSources interface
+// Will return next imageref string
+func (a *SourceList) Next() (string, error) {
+	if len(a.Sources) == 0 {
+		return "", nil
+	}
+	current := a.Sources
+	newList := append(current[:0], current[0+1:]...)
+	a.Sources = newList
+	return current[0], nil
+}
+
+// Resolve returns a possible image sources for a given image
+// Satisfies registrycontext.ImageResolver interface
+func (a *ICSPSources) Resolve(image string) (registryclient.ImageSources, error) {
+	sources, err := a.alternative(image)
```

Comment:
Nit: no need for separate method, I'd just embed the contents of `alternative` method in here.

Reply:
done, thanks